File: scripts/python/src/train.py
"""
Train models for credit card default prediction.
Uses same models as Rust project: XGBoost, CatBoost, LightGBM, Random Forest.

Usage: python src/train.py
"""

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
import xgboost as xgb
import catboost as cb
import lightgbm as lgb
import joblib
import json
import os
from datetime import datetime


def train_xgboost(X_train, y_train, X_val, y_val):
    """Train XGBoost model with early stopping."""
    print("\n--- Training XGBoost ---")
    
    # Calculate scale_pos_weight for imbalanced data
    scale_pos_weight = (y_train == 0).sum() / (y_train == 1).sum()
    print(f"Scale pos weight: {scale_pos_weight:.2f}")
    
    model = xgb.XGBClassifier(
        n_estimators=500,
        learning_rate=0.05,
        max_depth=6,
        min_child_weight=3,
        subsample=0.8,
        colsample_bytree=0.8,
        scale_pos_weight=scale_pos_weight,
        random_state=42,
        eval_metric='auc',
        early_stopping_rounds=50,
        n_jobs=-1
    )
    
    model.fit(
        X_train, y_train,
        eval_set=[(X_val, y_val)],
        verbose=50
    )
    
    print(f"✓ XGBoost trained ({model.best_iteration} rounds)")
    return model


def train_catboost(X_train, y_train, X_val, y_val):
    """Train CatBoost model with early stopping."""
    print("\n--- Training CatBoost ---")
    
    model = cb.CatBoostClassifier(
        iterations=500,
        learning_rate=0.05,
        depth=6,
        l2_leaf_reg=3,
        auto_class_weights='Balanced',
        random_seed=42,
        early_stopping_rounds=50,
        verbose=50
    )
    
    model.fit(
        X_train, y_train,
        eval_set=(X_val, y_val),
        use_best_model=True
    )
    
    print(f"✓ CatBoost trained ({model.best_iteration_} iterations)")
    return model


def train_lightgbm(X_train, y_train, X_val, y_val):
    """Train LightGBM model with early stopping."""
    print("\n--- Training LightGBM ---")
    
    model = lgb.LGBMClassifier(
        n_estimators=500,
        learning_rate=0.05,
        max_depth=6,
        num_leaves=31,
        min_child_samples=20,
        subsample=0.8,
        colsample_bytree=0.8,
        is_unbalance=True,
        random_state=42,
        n_jobs=-1
    )
    
    model.fit(
        X_train, y_train,
        eval_set=[(X_val, y_val)],
        callbacks=[
            lgb.early_stopping(stopping_rounds=50, verbose=True),
            lgb.log_evaluation(period=50)
        ]
    )
    
    print(f"✓ LightGBM trained ({model.best_iteration_} rounds)")
    return model


def train_random_forest(X_train, y_train):
    """Train Random Forest model."""
    print("\n--- Training Random Forest ---")
    
    model = RandomForestClassifier(
        n_estimators=200,
        max_depth=10,
        min_samples_split=5,
        min_samples_leaf=2,
        max_features='sqrt',
        class_weight='balanced',
        random_state=42,
        n_jobs=-1
    )
    
    model.fit(X_train, y_train)
    
    print(f"✓ Random Forest trained ({model.n_estimators} trees)")
    return model


def validate_model(model, X_val, y_val, model_name):
    """Validate a model and return ROC-AUC."""
    if hasattr(model, 'predict_proba'):
        y_proba = model.predict_proba(X_val)[:, 1]
    else:
        y_proba = model.predict(X_val)
    
    auc = roc_auc_score(y_val, y_proba)
    print(f"{model_name} Validation ROC-AUC: {auc:.4f}")
    return auc


def main():
    print("=" * 60)
    print("MODEL TRAINING")
    print("Models: XGBoost, CatBoost, LightGBM, Random Forest")
    print("=" * 60)
    
    # Load processed data
    print("\nLoading processed data...")
    X_train_full = pd.read_csv("results/X_train.csv")
    y_train_full = pd.read_csv("results/y_train.csv").values.ravel()
    
    print(f"Train samples: {len(X_train_full)}")
    print(f"Default rate: {y_train_full.mean():.1%}")
    print(f"Features: {len(X_train_full.columns)}")
    
    # Split into train and validation
    X_train, X_val, y_train, y_val = train_test_split(
        X_train_full, y_train_full,
        test_size=0.1,
        random_state=42,
        stratify=y_train_full
    )
    
    print(f"\nTraining: {len(X_train)} samples")
    print(f"Validation: {len(X_val)} samples")
    
    # Train all models
    models = {}
    
    # 1. XGBoost
    models['xgboost'] = train_xgboost(X_train, y_train, X_val, y_val)
    
    # 2. CatBoost
    models['catboost'] = train_catboost(X_train, y_train, X_val, y_val)
    
    # 3. LightGBM
    models['lightgbm'] = train_lightgbm(X_train, y_train, X_val, y_val)
    
    # 4. Random Forest
    models['random_forest'] = train_random_forest(X_train, y_train)
    
    # Validate all models
    print("\n" + "=" * 60)
    print("VALIDATION RESULTS")
    print("=" * 60)
    
    val_scores = {}
    for name, model in models.items():
        auc = validate_model(model, X_val, y_val, name)
        val_scores[name] = auc
    
    # Save models
    print("\n--- Saving Models ---")
    os.makedirs("results", exist_ok=True)
    
    for name, model in models.items():
        model_path = f"results/{name}_model.pkl"
        joblib.dump(model, model_path)
        print(f"✓ Saved: {model_path}")
    
    # Save validation scores
    val_df = pd.DataFrame([
        {'model': name, 'val_roc_auc': score}
        for name, score in val_scores.items()
    ]).sort_values('val_roc_auc', ascending=False)
    
    val_df.to_csv("results/validation_scores.csv", index=False)
    
    # Save feature names
    feature_names = list(X_train.columns)
    with open("results/feature_names.json", "w") as f:
        json.dump({
            'features': feature_names,
            'n_features': len(feature_names),
            'timestamp': datetime.now().isoformat()
        }, f, indent=2)
    
    print("\n✓ All models saved to results/")
    print(f"\nBest model: {val_df.iloc[0]['model']} (ROC-AUC: {val_df.iloc[0]['val_roc_auc']:.4f})")


if __name__ == "__main__":
    main()

File: scripts/r/src/train.R
# Train models for credit card default prediction
# Models: XGBoost, LightGBM, CatBoost (optional), Random Forest (same as Rust pipeline)
# Usage: Rscript src/train.R

library(tidyverse)
library(caret)
library(xgboost)
library(lightgbm)
library(randomForest)
library(pROC)

# Try to load CatBoost (optional - harder to install)
CATBOOST_AVAILABLE <- requireNamespace("catboost", quietly = TRUE)
if (CATBOOST_AVAILABLE) {
  library(catboost)
  cat("CatBoost available\n")
} else {
  cat("CatBoost not available (optional) - skipping\n")
}

# Feature groups
NUMERIC_FEATURES <- c(
  "LIMIT_BAL", "AGE",
  "BILL_AMT1", "BILL_AMT2", "BILL_AMT3", "BILL_AMT4", "BILL_AMT5", "BILL_AMT6",
  "PAY_AMT1", "PAY_AMT2", "PAY_AMT3", "PAY_AMT4", "PAY_AMT5", "PAY_AMT6",
  "avg_bill", "avg_payment", "payment_ratio", "credit_util"
)

ORDINAL_FEATURES <- c(
  "PAY_1", "PAY_2", "PAY_3", "PAY_4", "PAY_5", "PAY_6",
  "times_late", "max_months_late"
)

CATEGORICAL_FEATURES <- c("SEX", "EDUCATION", "MARRIAGE")

# Prepare data for modeling
prepare_data <- function(X, y = NULL) {
  # Select features
  features <- c(NUMERIC_FEATURES, ORDINAL_FEATURES, CATEGORICAL_FEATURES)
  features <- features[features %in% colnames(X)]
  
  X_prep <- X %>% select(all_of(features))
  
  # Convert to matrix for XGBoost/LightGBM
  X_matrix <- as.matrix(X_prep)
  
  if (!is.null(y)) {
    return(list(X = X_matrix, y = as.numeric(y), features = features, df = X_prep))
  }
  return(list(X = X_matrix, features = features, df = X_prep))
}

# Train XGBoost
train_xgboost <- function(X_train, y_train, X_val = NULL, y_val = NULL) {
  cat("\n--- Training XGBoost ---\n")
  
  # Calculate scale_pos_weight for class imbalance
  scale_pos_weight <- sum(y_train == 0) / sum(y_train == 1)
  cat("Scale pos weight:", round(scale_pos_weight, 2), "\n")
  
  dtrain <- xgb.DMatrix(data = X_train, label = y_train)
  
  params <- list(
    objective = "binary:logistic",
    eval_metric = "auc",
    max_depth = 6,
    eta = 0.1,
    subsample = 0.8,
    colsample_bytree = 0.8,
    scale_pos_weight = scale_pos_weight
  )
  
  watchlist <- list(train = dtrain)
  if (!is.null(X_val)) {
    dval <- xgb.DMatrix(data = X_val, label = y_val)
    watchlist <- list(train = dtrain, val = dval)
  }
  
  model <- xgb.train(
    params = params,
    data = dtrain,
    nrounds = 200,
    watchlist = watchlist,
    early_stopping_rounds = 50,
    verbose = 1,
    print_every_n = 50
  )
  
  cat("✓ XGBoost trained (", model$best_iteration, " rounds)\n", sep = "")
  return(model)
}

# Train LightGBM
train_lightgbm <- function(X_train, y_train, X_val = NULL, y_val = NULL) {
  cat("\n--- Training LightGBM ---\n")
  
  dtrain <- lgb.Dataset(data = X_train, label = y_train)
  
  params <- list(
    objective = "binary",
    metric = "auc",
    learning_rate = 0.1,
    max_depth = 6,
    num_leaves = 31,
    is_unbalance = TRUE,
    verbose = -1
  )
  
  valids <- list(train = dtrain)
  if (!is.null(X_val)) {
    dval <- lgb.Dataset(data = X_val, label = y_val, reference = dtrain)
    valids <- list(train = dtrain, val = dval)
  }
  
  model <- lgb.train(
    params = params,
    data = dtrain,
    nrounds = 200,
    valids = valids,
    early_stopping_rounds = 50,
    verbose = 1
  )
  
  cat("✓ LightGBM trained (", model$best_iter, " rounds)\n", sep = "")
  return(model)
}

# Train CatBoost
train_catboost <- function(X_train, y_train, X_val = NULL, y_val = NULL, feature_names = NULL) {
  cat("\n--- Training CatBoost ---\n")
  
  # Convert feature_names to list (CatBoost requirement)
  feat_list <- if (!is.null(feature_names)) as.list(feature_names) else NULL
  
  train_pool <- catboost.load_pool(
    data = X_train,
    label = y_train,
    feature_names = feat_list
  )
  
  params <- list(
    loss_function = "Logloss",
    eval_metric = "AUC",
    iterations = 200,
    learning_rate = 0.1,
    depth = 6,
    auto_class_weights = "Balanced",
    verbose = 50
  )
  
  if (!is.null(X_val)) {
    val_pool <- catboost.load_pool(
      data = X_val,
      label = y_val,
      feature_names = feat_list
    )
    model <- catboost.train(train_pool, val_pool, params = params)
  } else {
    model <- catboost.train(train_pool, params = params)
  }
  
  cat("✓ CatBoost trained\n")
  return(model)
}

# Train Random Forest
train_random_forest <- function(X_train, y_train) {
  cat("\n--- Training Random Forest ---\n")
  
  # Convert to data frame for randomForest
  df <- as.data.frame(X_train)
  df$target <- as.factor(y_train)
  
  model <- randomForest(
    target ~ .,
    data = df,
    ntree = 200,
    mtry = floor(sqrt(ncol(X_train))),
    classwt = c("0" = 1, "1" = sum(y_train == 0) / sum(y_train == 1)),
    importance = TRUE
  )
  
  cat("✓ Random Forest trained (", model$ntree, " trees)\n", sep = "")
  return(model)
}

# Main training function
main <- function() {
  cat("=== MODEL TRAINING ===\n")
  cat("Models: XGBoost, LightGBM, CatBoost, Random Forest\n\n")
  
  # Load data
  cat("Loading processed data...\n")
  X_train <- read_csv("results/X_train.csv", show_col_types = FALSE)
  y_train <- read_csv("results/y_train.csv", show_col_types = FALSE)$default
  
  cat("Train samples:", nrow(X_train), "\n")
  cat("Default rate:", round(mean(y_train) * 100, 1), "%\n")
  
  # Prepare data
  prep <- prepare_data(X_train, y_train)
  X_mat <- prep$X
  y_vec <- prep$y
  features <- prep$features
  
  cat("Features:", length(features), "\n\n")
  
  # Split for validation
  set.seed(42)
  val_idx <- sample(1:nrow(X_mat), size = floor(0.1 * nrow(X_mat)))
  X_train_split <- X_mat[-val_idx, ]
  y_train_split <- y_vec[-val_idx]
  X_val <- X_mat[val_idx, ]
  y_val <- y_vec[val_idx]
  
  cat("Training:", nrow(X_train_split), "samples\n")
  cat("Validation:", nrow(X_val), "samples\n")
  
  # Train models
  models <- list()
  
  # 1. XGBoost
  models$xgboost <- train_xgboost(X_train_split, y_train_split, X_val, y_val)
  
  # 2. LightGBM
  models$lightgbm <- train_lightgbm(X_train_split, y_train_split, X_val, y_val)
  
  # 3. CatBoost (if available)
  if (CATBOOST_AVAILABLE) {
    models$catboost <- train_catboost(X_train_split, y_train_split, X_val, y_val, features)
  } else {
    cat("\n--- Skipping CatBoost (not installed) ---\n")
  }
  
  # 4. Random Forest
  models$random_forest <- train_random_forest(X_train_split, y_train_split)
  
  # Save models
  cat("\n--- Saving Models ---\n")
  saveRDS(models, "results/models.rds")
  saveRDS(features, "results/features.rds")
  
  cat("\n✓ All models saved to results/models.rds\n")
  cat("✓ Feature names saved to results/features.rds\n")
}

# Run if executed directly
if (!interactive()) {
  main()
}

